src/sinks/datadog/events.rs

```diff
@@ -0,0 +1,444 @@
+        },
+        Healthcheck, VectorSink,
+    },
+    tls::{MaybeTlsSettings, TlsConfig},
+};
+use futures::{FutureExt, SinkExt};
+use http::Request;
+use indoc::indoc;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::{sync::Arc, time::Duration};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct DatadogEventsConfig {
+    endpoint: Option<String>,
+
+    #[serde(default = "default_site")]
+    site: String,
+    default_api_key: String,
+
+    tls: Option<TlsConfig>,
+
+    #[serde(default)]
+    request: TowerRequestConfig<Concurrency>,
+}
+
+type ApiKey = Arc<str>;
+
+fn default_site() -> String {
+    "datadoghq.com".to_owned()
+}
+
+inventory::submit! {
+    SinkDescription::new::<DatadogEventsConfig>("datadog_events")
+}
+
+impl GenerateConfig for DatadogEventsConfig {
+    fn generate_config() -> toml::Value {
+        toml::from_str(indoc! {r#"
+            default_api_key = "${DATADOG_API_KEY_ENV_VAR}"
+        "#})
+        .unwrap()
+    }
+}
+
+impl DatadogEventsConfig {
+    fn get_uri(&self) -> String {
+        format!("{}/api/v1/events", self.get_api_endpoint())
+    }
+
+    fn get_api_endpoint(&self) -> String {
+        self.endpoint
+            .clone()
+            .unwrap_or_else(|| format!("https://api.{}", &self.site))
+    }
+
+    fn build_sink<T, B, O>(
+        &self,
+        cx: SinkContext,
+        service: T,
+        batch: B,
+        timeout: Duration,
+    ) -> crate::Result<(VectorSink, Healthcheck)>
+    where
+        O: 'static,
+        B: Batch<Output = Vec<O>> + std::marker::Send + 'static,
+        B::Output: std::marker::Send + Clone,
+        B::Input: std::marker::Send,
+        T: HttpSink<
+                Input = PartitionInnerBuffer<B::Input, ApiKey>,
+                Output = PartitionInnerBuffer<B::Output, ApiKey>,
+            > + Clone,
+    {
+        let mut request_opts = self.request;
+        // Since we are sending only one event per request we should try to send as much
+        // requests in parallel as possible.
+        request_opts.concurrency = request_opts.concurrency.if_none(Concurrency::Adaptive);
+        let request_settings = request_opts.unwrap_with(&TowerRequestConfig::default());
+
+        let tls_settings = MaybeTlsSettings::from_config(
+            &Some(self.tls.clone().unwrap_or_else(TlsConfig::enabled)),
+            false,
+        )?;
+
+        let client = HttpClient::new(tls_settings)?;
+        let healthcheck = healthcheck(
+            self.get_api_endpoint(),
+            self.default_api_key.clone(),
+            client.clone(),
+        )
+        .boxed();
+        let sink = PartitionHttpSink::new(
+            service,
+            PartitionBuffer::new(batch),
+            request_settings,
+            timeout,
+            client,
+            cx.acker(),
+        )
+        .sink_map_err(|error| error!(message = "Fatal datadog_metrics sink error.", %error));
```

Comment:
ITYM `datadog_events`?

A lot of this appears to be identical to `datadog_logs` and `datadog_metrics`, suggesting an opportunity to share the implementation.

Reply:
> ITYM datadog_events?

Yep.

> A lot of this appears to be identical to datadog_logs and datadog_metric

They kinda are, but are just enough different for sharing not to be really useful. At best `build_sink` is almost identical and could be shared but that function would look quite arbitrary and with a bunch of variables. Also, Events API seems to be new with some fundamental things like compression and batching missing, once/if they add that at that point they will have more things to share.